src/hooks/api/authenticated-query-v2.ts

```diff
@@ -0,0 +1,162 @@
+import { NextApiRequest } from 'next';
+import { useRouter } from 'next/router';
+import { Cookies } from 'react-cookie';
+import {
+  FetchQueryOptions,
+  QueryClient,
+  QueryFunctionContext,
+  QueryKey,
+  useQuery,
+  UseQueryOptions,
+} from 'react-query';
+
+import { FetchError } from '@/utils/fetchFromApi';
+import { isTokenValid } from '@/utils/isTokenValid';
+
+import { useCookiesWithOptions } from '../useCookiesWithOptions';
+import { Headers } from './types/Headers';
+import { createHeaders, useHeaders } from './useHeaders';
+
+type QueryArguments = Record<string, string>;
+
+type GenerateQueryKeyArguments = {
+  queryKey: QueryKey;
+  queryArguments: QueryArguments;
+};
+
+type GeneratedQueryKey = readonly [QueryKey, QueryArguments];
+
+type AuthenticatedQueryFunctionContext = QueryFunctionContext<GeneratedQueryKey> & {
+  headers: Headers;
+};
+
+type ServerSideOptions = {
+  req: NextApiRequest;
+  queryClient: QueryClient;
+};
+
+type PrefetchAuthenticatedQueryOptions<TQueryFnData> = {
+  queryArguments?: QueryArguments;
+} & ServerSideOptions &
+  FetchQueryOptions<TQueryFnData, FetchError, TQueryFnData, QueryKey>;
+
+type UseAuthenticatedQueryOptions<TQueryFnData> = {
+  queryArguments?: QueryArguments;
+} & UseQueryOptions<TQueryFnData, FetchError, TQueryFnData, QueryKey>;
+
+const isUnAuthorized = (status: number) => [401, 403].includes(status);
+
+const generateQueryKey = ({
+  queryKey,
+  queryArguments,
+}: GenerateQueryKeyArguments): GeneratedQueryKey => {
+  if (Object.keys(queryArguments ?? {}).length > 0) {
+    return [queryKey, queryArguments];
+  }
+
+  return [queryKey, {}];
+};
+
+type GetPreparedOptionsArguments<TQueryFnData> = {
+  options: UseAuthenticatedQueryOptions<TQueryFnData>;
+  isTokenPresent: boolean;
+  headers: Headers;
+};
+
+const getPreparedOptions = <TQueryFnData = unknown>({
+  options,
+  isTokenPresent,
+  headers,
+}: GetPreparedOptionsArguments<TQueryFnData>): UseQueryOptions<
+  TQueryFnData,
+  FetchError,
+  TQueryFnData,
+  GeneratedQueryKey
+> => {
+  const { queryKey, queryArguments, queryFn, ...restOptions } = options;
+  const generatedQueryKey = generateQueryKey({
+    queryKey,
+    queryArguments,
+  });
+
+  const queryFunctionWithHeaders = async (
+    context: QueryFunctionContext<GeneratedQueryKey>,
+  ) => {
+    return await queryFn(
+      context,
+      // @ts-expect-error
```

Comment:
Can we remove this `//@ts-expect-error` ?

Reply:
I think we won't be able to remove this one, because react-query doesn't expect another argument it only known `context` which contains `queryKey` and `pageParam`